### PR TITLE
New feature:Add BadgerDB support

### PR DESCRIPTION
**Pull request by diluga**
New feature:Add BadgerDB support
  https://github.com/juicedata/juicefs/issues/1326

### CODE REVIEW

**Comment by SandyXSD** [comment]
@diluga Thank you for the contribution! Could you also add unit tests here: https://github.com/juicedata/juicefs/blob/main/pkg/meta/tkv_test.go

File: pkg/meta/tkv_badger.go [hunk 1]
```diff
@@ -0,0 +1,276 @@
+/*
+ * JuiceFS, Copyright 2021 Juicedata, Inc.
```
Comment (davies) [inline pkg/meta/tkv_badger.go:2]: 2022

File: pkg/meta/tkv_badger.go [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+	for _, key := range keys {
+		if err := tx.t.Delete(key); err != nil {
+			panic(err)
+		}
+	}
+}
+
+type badgerClient struct {
+	client *badger.DB
+	*sync.RWMutex
```
Comment (davies) [inline pkg/meta/tkv_badger.go:235]: sync.Mutex should be enough and faster

File: pkg/meta/tkv_badger.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+}
+
+func (c *badgerClient) reset(prefix []byte) error {
+	tx := c.client.NewTransaction(true)
+	defer tx.Discard()
+	it := tx.NewIterator(badger.DefaultIteratorOptions)
+	defer it.Close()
+	for it.Rewind(); it.Valid(); it.Next() {
+		if err := tx.Delete(it.Item().Key()); err != nil {
+			panic(err)
```
Comment (davies) [inline pkg/meta/tkv_badger.go:272]: return the `err`

File: pkg/meta/tkv_badger.go [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+		}
+	}
+	return nil
```
Comment (davies) [inline pkg/meta/tkv_badger.go:275]: commit the transaction

File: pkg/meta/tkv_badger.go [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+	client *badger.DB
+	*sync.RWMutex
+}
+
+func (c *badgerClient) name() string {
+	return "badger"
+}
+
+func (c *badgerClient) txn(f func(kvTxn) error) error {
+	c.Lock()
```
Comment (davies) [inline pkg/meta/tkv_badger.go:243]: The transaction should be thread-safe, we don't need the lock

File: pkg/meta/tkv_badger.go [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+
+func (tx *badgerTxn) set(key, value []byte) {
+	err := tx.t.Set(key, value)
+	if err == badger.ErrTxnTooBig {
+		logger.Infof("Current txn too big:%v.\n", tx)
+		if er := tx.t.Commit(); er != nil {
+			panic(er)
+		}
+		//tx.t = tx.c.client.NewTransaction(true)
+		if er := tx.t.Set(key, value); er != nil {
```
Comment (davies) [inline pkg/meta/tkv_badger.go:196]: err = tx.t.Set(key, value)

File: pkg/meta/tkv_badger.go [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+			panic(err)
+		}
+
+	}
+	if err != nil && err != badger.ErrTxnTooBig {
```
Comment (davies) [inline pkg/meta/tkv_badger.go:201]: err != nil

File: pkg/meta/tkv_badger.go [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+func (tx *badgerTxn) exist(prefix []byte) bool {
+	nextKey := tx.nextKey(prefix)
+	it, err := tx.Iter(prefix, nextKey)
+	if err != nil {
+		panic(err)
+	}
+	defer it.Close()
+	if it.Valid() {
+		item := it.Item()
+		if bytes.Compare(item.Key(), nextKey) >= 0 {
```
Comment (davies) [inline pkg/meta/tkv_badger.go:181]: Iter() should return any key less than `nextKey`, so we don't need this check, right?

File: pkg/meta/tkv_badger.go [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+			return false
+		}
+	}
+	return it.Valid()
+}
+
+func (tx *badgerTxn) set(key, value []byte) {
+	err := tx.t.Set(key, value)
+	if err == badger.ErrTxnTooBig {
+		logger.Infof("Current txn too big:%v.\n", tx)
```
[the remark this answers is not in the document]
Comment (diluga) [reply pkg/meta/tkv_badger.go:191]: https://github.com/dgraph-io/badger/issues/1242

File: pkg/meta/tkv_badger.go [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+		}
+	}
+
+	for i, key := range keys {
+		values[i] = ret[string(key)]
+	}
+	return values
+}
+
+func (tx *badgerTxn) Iter(startKey []byte, upperBound []byte) (*badger.Iterator, error) {
```
Comment (davies) [inline pkg/meta/tkv_badger.go:94]: upperBound is not used here

File: pkg/meta/tkv_badger.go [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+}
+
+func (tx *badgerTxn) gets(keys ...[]byte) [][]byte {
+	values := make([][]byte, len(keys))
+	ret := make(map[string][]byte)
+	for _, key := range keys {
+		item, err := tx.t.Get(key)
+		if err == badger.ErrKeyNotFound {
+			ret[string(key)] = []byte{}
+		}
```
Comment (davies) [inline pkg/meta/tkv_badger.go:75]: else

File: pkg/meta/tkv_badger.go [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+		if err != nil {
+			panic(err)
+		}
+		err = item.Value(func(val []byte) error {
+			ret[string(key)] = append([]byte{}, val...)
```
Comment (davies) [inline pkg/meta/tkv_badger.go:80]: append may allocate more memory than make(), it's better to use `make([]byte)` here.

File: pkg/meta/tkv_badger.go [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+	return value
+}
+
+func (tx *badgerTxn) gets(keys ...[]byte) [][]byte {
+	values := make([][]byte, len(keys))
+	ret := make(map[string][]byte)
+	for _, key := range keys {
+		item, err := tx.t.Get(key)
+		if err == badger.ErrKeyNotFound {
+			ret[string(key)] = []byte{}
```
Comment (davies) [inline pkg/meta/tkv_badger.go:74]: continue

File: pkg/meta/tkv_badger.go [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+	})
+	if err != nil {
+		panic(err)
+	}
+	return value
+}
+
+func (tx *badgerTxn) gets(keys ...[]byte) [][]byte {
+	values := make([][]byte, len(keys))
+	ret := make(map[string][]byte)
```
Comment (davies) [inline pkg/meta/tkv_badger.go:70]: We don't need the ret, put the result into `values` directly.

File: pkg/meta/tkv_badger.go [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+	}
+}
+
+func (tx *badgerTxn) get(key []byte) []byte {
+	item, err := tx.t.Get(key)
+	if err == badger.ErrKeyNotFound {
+		return nil
+	}
+	var value []byte
+	err = item.Value(func(val []byte) error {
```
Comment (davies) [inline pkg/meta/tkv_badger.go:58]: item.ValueCopy() ?

**Comment by davies** [comment]
#1340 will take over this one, thanks!

**Pull request closed without merging**